include/ofi_mr.h

```diff
@@ -133,7 +132,18 @@ int ofi_monitor_subscribe(struct ofi_notification_queue *nq,
 			  struct ofi_subscription *subscription);
 void ofi_monitor_unsubscribe(struct ofi_subscription *subscription);
 struct ofi_subscription *ofi_monitor_get_event(struct ofi_notification_queue *nq);
-
+/* Must call with nq->lock held */
+static inline void
+ofi_monitor_add_event_to_nq(struct ofi_subscription *subscription)
+{
+	FI_DBG(&core_prov, FI_LOG_MR,
+	       "Add event to NQ, context=%p, addr=%p, len=%"PRIu64" nq=%p\n",
+	       subscription, subscription->addr,
+	       subscription->len, subscription->nq);
+	if (dlist_empty(&subscription->entry))
+		dlist_insert_tail(&subscription->entry,
+				  &subscription->nq->list);
+}
```

Comment:
Given the rest of the code, it makes sense to include the locking as part of this function.

Reply:
Thanks. Done